Add benchmarking and perf improvements to Lisa.

pmclanahan opened the pull request:
This is a branch and PR for perf work for lisa.py.

@jgmize feel free to discuss this and send commits here.
I made this a branch on the moz repo for collaboration.

pmac committed:
Add benchmarking option to lisa.

pmac committed:
Use redis pipeline for all IP data storage.

jgmize commented on smithers/lisa.py:
```diff
@@ -159,9 +175,13 @@ def main():
-            statsd.gauge('queue.geoip', redis.llen(rkeys.IPLOGS))
+        if args.benchmark:
+            if not counter % 1000:
+                print counter
```
I know you're only printing out every 1000, but any print statements at all can have a surprising effect on performance.

jgmize commented on smithers/lisa.py:
```diff
@@ -159,9 +175,13 @@ def main():
+        else:
+            if counter >= 1000:
+                counter = 0
+                statsd.gauge('queue.geoip', redis.llen(rkeys.IPLOGS))
```
This isn't included in the benchmark, but could possibly have an effect on the performance.

jgmize commented:
r+wc

The redis pipelining is a huge improvement, and I don't see any issues running multiple lisa procs with this, which I believe is the most practical solution that we can implement in the time we have.  I believe we could get some significant benefits with python3+asyncio, but as discussed on irc, that unfortunately isn't a realistic option for us before launch.

pmac committed:
Execute redis pipeline every 1000th iteration.

pmclanahan commented:
If it'd be significantly better we could possibly run only lisa on py3....

pmac committed:
Convert Lisa's rate limitor to in memory LRU cache.

This means that the rate limiting won't be shared
between procs, but will be much faster.

It also changes in that it won't allow more of the
same IP until timeout seconds after the one that
put it over the edge. This should be okay, but something
to watch.

pmclanahan commented:
I think we're good to go here. Anything you'd like to see changed before merging @jgmize ?

jgmize commented on smithers/lisa.py:
```diff
@@ -159,9 +176,14 @@ def main():
+            if counter >= 1000:
+                pipe.execute()
+                counter = 0
+                statsd.gauge('queue.geoip', redis.llen(rkeys.IPLOGS))
```
For improved benchmarking accuracy, shouldn't we include this in the args.benchmark branch?

pmclanahan replied:
probably. I think we're done running benches for now. We can move this though.

pmclanahan commented:
Thanks. Merging. We can address further things in another PR.
